### PR TITLE
[Inductor][Reliability] Add runtime numeric check for pt2 Optimus in the pre grad pass

### PR DESCRIPTION
**Pull request by mengluy0125**
Summary: Titled

Test Plan:
# local reproduce
Patch ``icfg.fx_passes_numeric_check["pre_fx_passes"] = True"
```
buck2 run @mode/opt //scripts/jackiexu0313/pt2:local_model_with_pt2 -- --test_mode split_batch
```
P965217137

# MC candidates
### FIRST + CMF
f520754604
P1056796962
### ICVR
f520816217
P1056839342
### IG_CTR
f520819178
P1056903302
### MAI
f520823559
P1057712009
### AFOC
f520822438
P1057760058
### DPA
f520826815
P1057808574
### How the runtime numeric check to catch [SEVs](https://docs.google.com/document/d/1WOtlbgCBbmU1klK1LiGSO0lYf_7mtSP4nAdvhQHM0JE/edit#heading=h.k61fy2rhaijp)
bug fix diff: D51378532
### CMF+(FIRST)
f509587388
P1058305139
by running the numeric check, we can catch the forward loss differences (e.g., diffing(https://www.internalfb.com/intern/diffing/?paste_number=1058293804))
https://pxl.cl/4bQDG

f501760099
P1058400691
by running the numeric check, we can catch the forward loss differences (e.g., diffing(https://www.internalfb.com/intern/diffing/?paste_number=1058412054))
https://pxl.cl/4bQMw

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

File: torch/_inductor/fx_passes/numeric_utils.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+        res = compare_parameters(model_base, model_control, precision)
+        logger.info("compare parameters. Numerical result : %s", res)
+
+        res = compare_forward_output(pred_base, pred_control, precision)
+        logger.info("compare loss/predict. Numerical result : %s", res)
+        # tensor may not have a grad_fn
+        try:
+            _ = pred_base[0].sum().backward(retain_graph=True)
+            _ = pred_control[0].sum().backward(retain_graph=True)
```
Comment (jackiexu1992) [inline torch/_inductor/fx_passes/numeric_utils.py:144]:
```suggestion
            _ = pred_control["loss"].sum().backward()
```
but this could be internal model specific.

File: torch/_inductor/fx_passes/numeric_utils.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+# If the two forward functions involve any non-deterministic operations,
+# such as certain types of parallelism or asynchronous execution,
+# this can also lead to different outputs.
+def set_deterministic() -> None:
+    """Make torch manual seed deterministic."""
+    import random
+
+    import numpy
+
+    import torch
```
[the remark this answers is not in the document]
Comment (mengluy0125) [reply torch/_inductor/fx_passes/numeric_utils.py:27]: Sure, moved :).

File: torch/_inductor/fx_passes/numeric_utils.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    if len(tuple_base) != len(tuple_control):
+        logger.warning(
+            "Mismatch fw output length. pt2.length: %s, pt2-optimus.length: %s",
+            len(tuple_base),
+            len(tuple_control),
+        )
+        return False
+    is_allclose = True
+    for i in range(len(tuple_base)):
+        # The default model precision is ft32, unless users manually set it to fp16.
```
[the remark this answers is not in the document]
Comment (mengluy0125) [reply torch/_inductor/fx_passes/numeric_utils.py:84]: We support user configuration to change the precision in case they want to have a more strict numerical check.

File: torch/_inductor/fx_passes/numeric_utils.py (hunk 1 above)
Comment (mengluy0125) [reply torch/_inductor/fx_passes/numeric_utils.py:144]: The output is a tuple, we need to use 0 index.

**Review by jackiexu1992 — APPROVED** [review]

File: torch/_inductor/fx_passes/pre_grad.py [hunk 4]
```diff
@@ -69,6 +71,8 @@ def pre_grad_passes(gm: torch.fx.GraphModule, example_inputs):
 
     if config.pattern_matcher:
         lazy_init()
+        if config.runtime_numeric_check["enable"]:
+            pre_optimus = gm.__copy__()
```
Comment (yanboliang) [inline torch/_inductor/fx_passes/pre_grad.py:75]: It's better to name as ```pre_fx_passes``` since this is a useful tool for all fx passes.

File: torch/_inductor/fx_passes/pre_grad.py [hunk 5]
```diff
@@ -92,6 +96,25 @@ def pre_grad_passes(gm: torch.fx.GraphModule, example_inputs):
     gm.graph.lint()
     gm.recompile()
 
+    # need to topo-sort graphmodule before we run the model,
+    # otherwise it may fail as refer before def
+    if config.pattern_matcher and config.runtime_numeric_check["enable"]:
```
Comment (yanboliang) [inline torch/_inductor/fx_passes/pre_grad.py:101]: Let's refactor L99 to L117 into a function called ```numeric_check_if_enabled(gm_before_fx_passes, gm_after_fx_passes)``` and move the definition into ```numeric_utils.py```. I think that would make this file more clear logically.

File: torch/_inductor/config.py [hunk 6]
```diff
@@ -134,6 +134,18 @@ def is_fbcode():
 # Autotune will compare perf with normal cast->then->mm option
 use_mixed_mm = False
 
+# enable runtime numeric check for pre/post grad fx passes
+# floating point provides limited accuracy (about 7 decimal digits for single precision
+# floating point numbers,about 16 decimal digits for double precision floating point numbers)
+# according to PyTorch documentation.
+# https://pytorch.org/docs/stable/notes/numerical_accuracy.html#batched-computations-or-slice-computations
+fx_passes_numeric_check: Dict[str, Any] = {
+    "pre_fx_passes": False,
```
Comment (yanboliang) [inline torch/_inductor/config.py:143]: It's better to use ```pre_grad```, ```pre_fx_passes``` means check did before these fx passes.

**Review by yanboliang — APPROVED** [review]
